pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -1,95 +1,275 @@
 package upgrade
 
 import (
+	"context"
+	"fmt"
+	"sync"
+
+	provisioningv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
+	rkev1 "github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1"
+	provisioningctrl "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
 	v1 "github.com/rancher/wrangler/pkg/generated/controllers/batch/v1"
 	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 
 	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
 	ctlharvesterv1 "github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+	kubevirtctrl "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
 	upgradectlv1 "github.com/harvester/harvester/pkg/generated/controllers/upgrade.cattle.io/v1"
 	"github.com/harvester/harvester/pkg/settings"
 )
 
+var (
+	upgradeControllerLock sync.Mutex
+)
+
 const (
 	//system upgrade controller is deployed in cattle-system namespace
-	upgradeNamespace               = "cattle-system"
+	upgradeNamespace               = "harvester-system"
+	sucNamespace                   = "cattle-system"
 	upgradeServiceAccount          = "system-upgrade-controller"
 	harvesterSystemNamespace       = "harvester-system"
 	harvesterVersionLabel          = "harvesterhci.io/version"
 	harvesterUpgradeLabel          = "harvesterhci.io/upgrade"
 	harvesterManagedLabel          = "harvesterhci.io/managed"
 	harvesterLatestUpgradeLabel    = "harvesterhci.io/latestUpgrade"
 	harvesterUpgradeComponentLabel = "harvesterhci.io/upgradeComponent"
-	upgradeImageRepository         = "rancher/harvester-bundle"
+	harvesterNodeLabel             = "harvesterhci.io/node"
+	upgradeImageRepository         = "rancher/harvester-upgrade"
+
+	harvesterNodePendingOSImage     = "harvesterhci.io/pendingOSImage"
+	harvesterUpgradeImageAnnotation = "harvesterhci.io/upgradeImage"
+	repoInfoAnnotation              = "harvesterhci.io/repoInfo"
+	singleNodeAnnotation            = "harvesterhci.io/singleNode"
```

Comment:
Is it better to set them in upgrade status other than in upgrade annotations

Reply:
https://github.com/harvester/harvester/pull/1635/commits/2adcd6ba0ddf8041f42a4b0cd191021fb7881808
Fixed. I moved all annotation hacks into status.